gradle-witchcraft-logging/src/main/groovy/com/palantir/witchcraft/java/logging/gradle/testreport/FormattingTestReporter.java

```diff
@@ -69,30 +84,58 @@ private static final class FormattingTestResultsProvider implements TestResultsP
         }
 
         @Override
-        @SuppressWarnings("StringSplitter")
         public void writeAllOutput(long classId, TestOutputEvent.Destination destination, Writer writer) {
             if (destination == TestOutputEvent.Destination.StdErr
                     || destination == TestOutputEvent.Destination.StdOut) {
-                StringWriter stringWriter = new StringWriter();
-                delegate.writeAllOutput(classId, destination, stringWriter);
-                String contents = stringWriter.toString();
-                for (String line : contents.split("\n")) {
-                    try {
-                        PARSER.tryParse(line)
-                                .orElseGet(() -> out -> {
-                                    out.write(line);
-                                    out.write("\n");
-                                })
-                                .write(writer);
-                    } catch (IOException e) {
-                        throw new RuntimeException(e);
-                    }
-                }
+                delegate.writeAllOutput(classId, destination, new LineProcessingWriter(writer, LINE_PROCESSOR));
             } else {
                 delegate.writeAllOutput(classId, destination, writer);
             }
         }
 
+        private static class LineProcessingWriter extends Writer {
+            private final Writer delegate;
+            private final BiConsumer<String, Writer> lineProcessor;
+            private final StringBuilder lineBuffer = new StringBuilder();
+
+            LineProcessingWriter(Writer delegate, BiConsumer<String, Writer> lineProcessor) {
+                this.delegate = delegate;
+                this.lineProcessor = lineProcessor;
+            }
+
+            @Override
+            public void write(char[] cbuf, int off, int len) throws IOException {
+                for (int i = off; i < off + len; i++) {
+                    char ch = cbuf[i];
+                    if (ch == '\n') {
+                        processLine();
+                    } else {
+                        lineBuffer.append(ch);
+                    }
+                }
+            }
+
+            @Override
+            public void flush() throws IOException {
+                delegate.flush();
+            }
+
+            @Override
+            public void close() throws IOException {
+                if (!lineBuffer.isEmpty()) {
+                    processLine();
+                }
+                delegate.close();
+            }
+
+            private void processLine() throws IOException {
+                String line = lineBuffer.toString();
+                lineProcessor.accept(line, delegate);
+                delegate.write("\n");
```

Comment:
Do we need to synchronize on the writer to ensure this is written atomically?

```suggestion
                synchronized (this) {
                    lineProcessor.accept(line, delegate);
                    delegate.write('\n');
                }
```

Reply:
if we're going to make this LineProcessingWriter thread-safe, then I think we need to protect the lineBuffer from concurrent writes, and maybe then add `synchronized` keyword to the `write()` and `close()` methods.

I'm not sure we need it to be thread-safe though...  even multiple calls to the `writeAllOutput` method here would create new `LineProcessingWriter`s

Reply:
`java.io.Writer` exposes a `protected` `lock` field for synchronization to ensure that output (e.g. line with new line) is written as an atomic unit. I think something like https://github.com/palantir/witchcraft-java-logging/pull/763 would help here in cases where there might be concurrent things writing to stdout/stderr. I think we can also provide an optimized `write(String str, int off, int len)` that uses vectorized `String#indexOf` and bulk char array copy from `String` to `StringBuilder`.